src/nvim/fold.c

```diff
@@ -2228,32 +2228,51 @@ static linenr_T foldUpdateIEMSRecurse(garray_T *gap, int level,
              * before where we started looking, extend it.  If it
              * starts at another line, update nested folds to keep
              * their position, compensating for the new fd_top. */
-            if (fp->fd_top >= startlnum && fp->fd_top != firstlnum) {
-              if (fp->fd_top > firstlnum)
-                /* like lines are inserted */
+            if (fp->fd_top == firstlnum) {
+              // We have found a fold beginning exactly where we want one.
+            } else if (fp->fd_top >= startlnum) {
```

Comment:
I was splitting the `if/else if` clauses into logical sections:

If we've found a fold where we want it we don't have to do anything,
else if the fold starts below the update range ...
else if we are starting a new fold ...

before, the check that the fold doesn't start where we want it was combined with the other checks, which I thought made reading it less clear.

if the fold starts below the update range, and doesn't start where we want it ...
else if we are starting a new fold, and the one we've found doesn't start where we want it ...